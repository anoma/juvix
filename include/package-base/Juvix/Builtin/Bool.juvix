module Juvix.Builtin.Bool;

--- Inductive definition of booleans.
builtin bool
type Bool :=
  | true
  | false;
