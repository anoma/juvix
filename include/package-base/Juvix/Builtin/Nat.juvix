module Juvix.Builtin.Nat;

--- Inductive natural numbers. I.e. whole non-negative numbers.
builtin nat
type Nat :=
  | zero
  | suc Nat;
