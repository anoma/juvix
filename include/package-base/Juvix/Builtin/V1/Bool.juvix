module Juvix.Builtin.V1.Bool;

--- Inductive definition of booleans.
builtin bool
type Bool :=
  | false
  | true;

open Bool using {false; true} public;
