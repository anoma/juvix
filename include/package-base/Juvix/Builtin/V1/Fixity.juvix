module Juvix.Builtin.V1.Fixity;

syntax fixity none := none;

syntax fixity rapp := binary {assoc := right};
syntax fixity lapp := binary {assoc := left; same := rapp};
syntax fixity seq := binary {assoc := left; above := [lapp]};

syntax fixity functor := binary {assoc := right};

syntax fixity logical := binary {assoc := right; above := [seq]};
syntax fixity comparison := binary {assoc := none; above := [logical]};

syntax fixity pair := binary {assoc := right};
syntax fixity cons := binary {assoc := right; above := [pair]};

syntax fixity step := binary {assoc := right};
syntax fixity range := binary {assoc := right; above := [step]};

syntax fixity additive := binary {assoc := left; above := [comparison; range; cons]};
syntax fixity multiplicative := binary {assoc := left; above := [additive]};

syntax fixity composition := binary {assoc := right; above := [multiplicative]};
syntax fixity lcomposition := binary {assoc := left; above := [multiplicative]};
