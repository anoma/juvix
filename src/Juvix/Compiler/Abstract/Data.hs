module Juvix.Compiler.Abstract.Data
  ( module Juvix.Compiler.Abstract.Data.Name,
    module Juvix.Compiler.Abstract.Data.InfoTable,
    module Juvix.Compiler.Abstract.Data.InfoTableBuilder,
  )
where

import Juvix.Compiler.Abstract.Data.InfoTable
import Juvix.Compiler.Abstract.Data.InfoTableBuilder
import Juvix.Compiler.Abstract.Data.Name
