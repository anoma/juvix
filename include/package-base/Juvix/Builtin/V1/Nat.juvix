module Juvix.Builtin.V1.Nat;

import Juvix.Builtin.V1.Trait.Natural open public;
import Juvix.Builtin.V1.Nat.Base open hiding {+; *; div; mod} public;
import Juvix.Builtin.V1.Nat.Base as Nat;

{-# specialize: true, inline: case #-}
instance
naturalNatI : Natural Nat :=
  mkNatural@{
    + := (Nat.+);
    * := (Nat.*);
    div := Nat.div;
    mod := Nat.mod;
    fromNat (x : Nat) : Nat := x
  };
