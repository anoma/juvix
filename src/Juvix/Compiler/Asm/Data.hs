module Juvix.Compiler.Asm.Data
  ( module Juvix.Compiler.Asm.Data.InfoTable,
    module Juvix.Compiler.Asm.Data.InfoTableBuilder,
    module Juvix.Compiler.Asm.Data.Stack,
  )
where

import Juvix.Compiler.Asm.Data.InfoTable
import Juvix.Compiler.Asm.Data.InfoTableBuilder
import Juvix.Compiler.Asm.Data.Stack hiding (empty)
