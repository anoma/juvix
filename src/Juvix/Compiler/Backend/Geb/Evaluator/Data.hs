module Juvix.Compiler.Backend.Geb.Evaluator.Data
  ( module Juvix.Compiler.Backend.Geb.Evaluator.Data.Env,
    module Juvix.Compiler.Backend.Geb.Evaluator.Data.Values,
    module Juvix.Compiler.Backend.Geb.Evaluator.Data.RunEvalResult,
  )
where

import Juvix.Compiler.Backend.Geb.Evaluator.Data.Env
import Juvix.Compiler.Backend.Geb.Evaluator.Data.RunEvalResult
import Juvix.Compiler.Backend.Geb.Evaluator.Data.Values
