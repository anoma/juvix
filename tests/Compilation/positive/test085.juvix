-- Deriving Eq
module test085;

import Stdlib.Data.Nat open hiding {isEqual};
import Stdlib.Data.Fixity open;
import Stdlib.Data.Bool open;
import Stdlib.Data.Pair open;
import Stdlib.Trait.Eq open;
import Stdlib.System.IO open;

syntax alias isEqual := Eq.isEqual;

type NTree :=
  | NLeaf
  | NNode NTree Nat NTree;

open NTree using {NLeaf; NNode} public;

deriving instance
ntreeEq : Eq NTree;

type Tree A :=
  | Leaf
  | Node (Tree A) A (Tree A);

open Tree using {Leaf; Node} public;

deriving instance
treeEqI {A} {{Eq A}} : Eq (Tree A);

t1 : Tree Nat := Node (Node Leaf 0 Leaf) 1 Leaf;

t2 : Tree Nat := Node (Node Leaf 1 Leaf) 1 Leaf;

t3 : Tree Nat := Node Leaf 1 (Node Leaf 0 Leaf);

main : IO :=
  printLn (isEqual true true)
    >>> printLn (not (isEqual false true))
    >>> printLn (isEqual (false, true) (false, true))
    >>> printLn (not (isEqual (false, true) (false, false)))
    >>> printLn (isEqual t1 t1)
    >>> printLn (isEqual t1 t2)
    >>> printLn (isEqual t1 t3)
    >>> printLn (isEqual (NNode NLeaf 0 NLeaf) (NNode NLeaf 0 NLeaf));
