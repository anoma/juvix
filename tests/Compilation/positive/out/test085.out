true
true
true
true
true
false
false
